vowpalwabbit/example.h

```diff
@@ -52,6 +52,8 @@ typedef union
 { float scalar;
   uint32_t multiclass;
   MULTILABEL::labels multilabels;
+  float* probs; // for --probabilities --oaa
```

Comment:
If things are to work under reduction (we want them to!), you need to have
allocated memory associated with the example rather than the reduction.
Otherwise, a reduction which calls oaa twice will get corrupted results.

-John


On Thu, Oct 29, 2015 at 9:21 AM, Martin Popel <notifications@github.com>
wrote:

> In vowpalwabbit/example.h
> <https://github.com/JohnLangford/vowpal_wabbit/pull/837#discussion_r43382952>
> :
>
> > @@ -52,6 +52,8 @@ typedef union
> >  { float scalar;
> >    uint32_t multiclass;
> >    MULTILABEL::labels multilabels;
> > +  float* probs; // for --probabilities --oaa
>
> This is just a pointer. In oaa.cc, there is both data.probs =
> calloc_or_throw<float>(data.k) and data.probs =
> calloc_or_throw<float>(data.k). So I hope there are no memory leaks.
>
> Why not using v_array? I don't need dynamically growing array. The size
> is fixed to k (number of classes).
>
> Why not allocating and freeing probs within each example? First, I don't
> see a use case for this (where I would need to access probs of all the
> ring_size examples *simultaneously*). Second, I think allocating new
> memory for each example is slower than reusing one array for all the
> examples.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JohnLangford/vowpal_wabbit/pull/837/files#r43382952>.
>